In [1]:
# Dependencies and Setup
from pathlib import Path
import pandas as pd

In [2]:
# File to Load
school_data_to_load = Path("/Users/lisashimano/Documents/UWA Data Analytics Bootcamp/Week 4/Challenge/Starter_Code 2/PyCitySchools/Resources/schools_complete.csv")
student_data_to_load = Path("/Users/lisashimano/Documents/UWA Data Analytics Bootcamp/Week 4/Challenge/Starter_Code 2/PyCitySchools/Resources/students_complete.csv")

In [3]:
# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

In [4]:
# Combine the data into a single dataset.
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

In [5]:
# Calculate the Totals (Schools and Students)
school_count = len(school_data)
student_count = len(student_data)
print("Total number of unique schools:", school_count)
print("Total students:", student_count)

Total number of unique schools: 15
Total students: 39170


In [6]:
# Calculate the Total Budget
total_budget = school_data['budget'].sum()
print("Total budget:", total_budget)

Total budget: 24649428


In [7]:
# Calculate the Average Scores
average_maths_score = student_data['maths_score'].mean()
average_reading_score = student_data['reading_score'].mean()
print("Average maths score:", average_maths_score)
print("Average reading score:", average_reading_score)

Average maths score: 70.33819249425581
Average reading score: 69.98013786060761


In [8]:
# Calculate the Percentage Pass Rates
passing_maths_threshold = 50
passing_maths_count = student_data[(student_data["maths_score"] >= 50)].count()["student_name"]
passing_maths_percentage = passing_maths_count / float(student_count) * 100
passing_reading_threshold = 50
passing_reading_count = student_data[(student_data["reading_score"] >= 50)].count()["student_name"]
passing_reading_percentage = passing_reading_count / float(student_count) * 100
passing_maths_reading_df = student_data[(student_data["maths_score"] >= 50) & (student_data["reading_score"] >= 50)]
passing_maths_reading_count = len(passing_maths_reading_df)
overall_passing_rate = float(passing_maths_reading_count) / student_count * 100
print("% of students passing maths:", passing_maths_percentage)
print("% of students passing reading:", passing_reading_percentage)
print("% of students passing both maths and reading:", overall_passing_rate)

% of students passing maths: 86.07863160582077
% of students passing reading: 84.42685728874139
% of students passing both maths and reading: 72.80827163645647


In [9]:
# Convert to DataFrame
area_summary = pd.DataFrame({
    'Total Schools': [school_count],
    'Total Students': [student_count],
    'Total Budget': [total_budget],
    'Average Maths Score': [average_maths_score],
    'Average Reading Score': [average_reading_score],
    '% Passing Maths': [passing_maths_percentage],
    '% Passing Reading': [passing_reading_percentage],
    '% Overall Passing': [overall_passing_rate]
})

# Formatting
area_summary["Total Students"] = area_summary["Total Students"].map("{:,}".format)
area_summary["Total Budget"] = area_summary["Total Budget"].map("${:,.2f}".format)

#Create a DataFrame
df = pd.DataFrame(area_summary, columns=["Total Schools", "Total Students", "Total Budget", "Average Maths Score", "Average Reading Score", "% Passing Maths", "% Passing Reading", "% Overall Passing"])

# Display the DataFrame
area_summary.head()

Total Schools Total Students    Total Budget  Average Maths Score  \
0             15         39,170  $24,649,428.00            70.338192   

   Average Reading Score  % Passing Maths  % Passing Reading  \
0              69.980138        86.078632          84.426857   

   % Overall Passing  
0          72.808272

In [10]:
# Use the code provided to select the type per school from school_data
school_types = school_data.set_index(["school_name"])["type"]

# Calculate the total student count per school from school_data
per_school_counts = school_data.set_index(["school_name"])["size"]

# Calculate the total school budget and per capita spending per school from school_data
per_school_budget = school_data.set_index(["school_name"])["budget"]
per_school_capita = per_school_budget / per_school_counts

# Calculate the average test scores per school from school_data_complete
per_school_maths = school_data_complete.groupby("school_name")["maths_score"].mean()
per_school_reading = school_data_complete.groupby("school_name")["reading_score"].mean()

#Print the results
print("School Types:", school_types)
print("\nTotal Student Count per School:", per_school_counts)
print("\nTotal School Budget per School:", per_school_budget)
print("\nTotal Student Spending per School:", per_school_capita)
print("\nAverage Maths Score per School:", per_school_maths)
print("\nAverage Reading Score per School:", per_school_reading)

School Types: school_name
Huang High School         Government
Figueroa High School      Government
Shelton High School      Independent
Hernandez High School     Government
Griffin High School      Independent
Wilson High School       Independent
Cabrera High School      Independent
Bailey High School        Government
Holden High School       Independent
Pena High School         Independent
Wright High School       Independent
Rodriguez High School     Government
Johnson High School       Government
Ford High School          Government
Thomas High School       Independent
Name: type, dtype: object

Total Student Count per School: school_name
Huang High School        2917
Figueroa High School     2949
Shelton High School      1761
Hernandez High School    4635
Griffin High School      1468
Wilson High School       2283
Cabrera High School      1858
Bailey High School       4976
Holden High School        427
Pena High School          962
Wright High School       1800
Rodriguez High Sch

In [11]:
# Get the students who passed maths and passed reading by creating separate filtered DataFrames from school_data_complete.
school_passing_maths = school_data_complete[school_data_complete["maths_score"] >= 50]
school_passing_reading = school_data_complete[school_data_complete["reading_score"] >= 50]

# Get the students who passed both reading and maths in a separate DataFrame from school_data_complete.
passing_maths_and_reading = school_data_complete[(school_data_complete["maths_score"] >= 50) & (school_data_complete["reading_score"] >= 50)]

In [12]:
# Calculate the Percentage Pass Rates
per_school_passing_maths = school_passing_maths.groupby("school_name").size()
per_school_passing_maths_percent = (per_school_passing_maths / per_school_counts) * 100
per_school_passing_reading = school_passing_reading.groupby("school_name").size()
per_school_passing_reading_percent = (per_school_passing_reading / per_school_counts) * 100
per_school_passing_both = passing_maths_and_reading.groupby("school_name").size()
overall_passing_rate = (per_school_passing_both / per_school_counts) * 100

# Print the results
print("% of Students Passing Maths per School:", per_school_passing_maths_percent)
print("% of Students Passing Reading per School:", per_school_passing_reading_percent)
print("Overall Passing Rate per School:", overall_passing_rate)

% of Students Passing Maths per School: school_name
Bailey High School       91.639871
Cabrera High School      90.850377
Figueroa High School     81.654798
Ford High School         82.438846
Griffin High School      91.212534
Hernandez High School    80.949299
Holden High School       89.929742
Huang High School        81.693521
Johnson High School      82.062592
Pena High School         91.683992
Rodriguez High School    90.797699
Shelton High School      91.538898
Thomas High School       83.853211
Wilson High School       82.785808
Wright High School       91.777778
dtype: float64
% of Students Passing Reading per School: school_name
Bailey High School       87.379421
Cabrera High School      89.074273
Figueroa High School     82.807731
Ford High School         82.219788
Griffin High School      88.487738
Hernandez High School    81.877023
Holden High School       88.524590
Huang High School        81.453548
Johnson High School      81.978576
Pena High School         86.590437
Rodr

In [13]:
# Convert to DataFrame
per_school_summary = pd.DataFrame({
    "School Type": school_types,
    "Total Students": per_school_counts,
    "Total School Budget": per_school_budget,
    "Per Student Budget": per_school_capita,
    "Average Maths Score": per_school_maths,
    "Average Reading Score": per_school_reading,
    "% Passing Maths": per_school_passing_maths_percent,
    "% Passing Reading": per_school_passing_reading_percent,
    "% Overall Passing": overall_passing_rate
})
    
# Formatting
per_school_summary["Total School Budget"] = per_school_summary["Total School Budget"].map("${:,.2f}".format)
per_school_summary["Per Student Budget"] = per_school_summary["Per Student Budget"].map("${:,.2f}".format)

# Create a DataFrame
df = pd.DataFrame(per_school_summary, columns=["School Type", "Total Students", "Total School Budget", "Per Student Budget", "Average Maths Score", "Average Reading Score", "% Passing Maths", "% Passing Reading", "% Overall Passing"])

# Set display options to show all rows
pd.set_option('display.max_rows', None)

# Display the DataFrame
per_school_summary.head(15)

School Type  Total Students Total School Budget  \
school_name                                                              
Bailey High School      Government            4976       $3,124,928.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Figueroa High School    Government            2949       $1,884,411.00   
Ford High School        Government            2739       $1,763,916.00   
Griffin High School    Independent            1468         $917,500.00   
Hernandez High School   Government            4635       $3,022,020.00   
Holden High School     Independent             427         $248,087.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Pena High School       Independent             962         $585,858.00   
Rodriguez High School   Government            3999       $2,547,363.00   
Shelton High School    Independent            1761       $1,056,600.00   
Thomas High School     Independent            1635       $1,043,130.00   
Wilson High School     Independent            2283       $1,319,574.00   
Wright High School     Independent            1800       $1,049,400.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Bailey High School               $628.00            72.352894   
Cabrera High School              $582.00            71.657158   
Figueroa High School             $639.00            68.698542   
Ford High School                 $644.00            69.091274   
Griffin High School              $625.00            71.788147   
Hernandez High School            $652.00            68.874865   
Holden High School               $581.00            72.583138   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Pena High School                 $609.00            72.088358   
Rodriguez High School            $637.00            72.047762   
Shelton High School              $600.00            72.034072   
Thomas High School               $638.00            69.581651   
Wilson High School               $578.00            69.170828   
Wright High School               $583.00            72.047222   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Bailey High School                 71.008842        91.639871   
Cabrera High School                71.359526        90.850377   
Figueroa High School               69.077993        81.654798   
Ford High School                   69.572472        82.438846   
Griffin High School                71.245232        91.212534   
Hernandez High School              69.186408        80.949299   
Holden High School                 71.660422        89.929742   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Pena High School                   71.613306        91.683992   
Rodriguez High School              70.935984        90.797699   
Shelton High School                70.257808        91.538898   
Thomas High School                 69.768807        83.853211   
Wilson High School                 68.876916        82.785808   
Wright High School                 70.969444        91.777778   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Bailey High School             87.379421          80.084405  
Cabrera High School            89.074273          80.785791  
Figueroa High School           82.807731          67.650051  
Ford High School               82.219788          67.469880  
Griffin High School            88.487738          81.335150  
Hernandez High School          81.877023          66.364617  
Holden High School             88.524590          78.922

In [14]:
# Sort and show top five schools
school_summary_data = {
    "School Type": school_types,
    "Total Students": per_school_counts,
    "Total School Budget": per_school_budget,
    "Per Student Budget": per_school_capita,
    "Average Maths Score": per_school_maths,
    "Average Reading Score": per_school_reading,
    "% Passing Maths": per_school_passing_maths_percent,
    "% Passing Reading": per_school_passing_reading_percent,
    "% Overall Passing": overall_passing_rate
}

# Convert the dictionary to a DataFrame
school_summary_df = pd.DataFrame(school_summary_data)

# Formatting
school_summary_df["Total School Budget"] = school_summary_df["Total School Budget"].map("${:,.2f}".format)
school_summary_df["Per Student Budget"] = school_summary_df["Per Student Budget"].map("${:,.2f}".format)

# Sort the DataFrame by '% Overall Passing' in descending order
top_schools = school_summary_df.sort_values(by='% Overall Passing', ascending=False)

# Display the top 5 rows
print("Top 5 Schools by % Overall Passing:")
print(top_schools.head())

top_schools.head(5)

Top 5 Schools by % Overall Passing:
                       School Type  Total Students Total School Budget  \
school_name                                                              
Griffin High School    Independent            1468         $917,500.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Bailey High School      Government            4976       $3,124,928.00   
Wright High School     Independent            1800       $1,049,400.00   
Rodriguez High School   Government            3999       $2,547,363.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Griffin High School              $625.00            71.788147   
Cabrera High School              $582.00            71.657158   
Bailey High School               $628.00            72.352894   
Wright High School               $583.00            72.047222   
Rodriguez High School            $637.00            72.

School Type  Total Students Total School Budget  \
school_name                                                              
Griffin High School    Independent            1468         $917,500.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Bailey High School      Government            4976       $3,124,928.00   
Wright High School     Independent            1800       $1,049,400.00   
Rodriguez High School   Government            3999       $2,547,363.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Griffin High School              $625.00            71.788147   
Cabrera High School              $582.00            71.657158   
Bailey High School               $628.00            72.352894   
Wright High School               $583.00            72.047222   
Rodriguez High School            $637.00            72.047762   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Griffin High School                71.245232        91.212534   
Cabrera High School                71.359526        90.850377   
Bailey High School                 71.008842        91.639871   
Wright High School                 70.969444        91.777778   
Rodriguez High School              70.935984        90.797699   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Griffin High School            88.487738          81.335150  
Cabrera High School            89.074273          80.785791  
Bailey High School             87.379421          80.084405  
Wright High School             86.666667          79.722222  
Rodriguez High School          87.396849          79.419855

In [15]:
# Sort and show bottom five schools
school_summary_data = {
    "School Type": school_types,
    "Total Students": per_school_counts,
    "Total School Budget": per_school_budget,
    "Per Student Budget": per_school_capita,
    "Average Maths Score": per_school_maths,
    "Average Reading Score": per_school_reading,
    "% Passing Maths": per_school_passing_maths_percent,
    "% Passing Reading": per_school_passing_reading_percent,
    "% Overall Passing": overall_passing_rate
}

# Convert the dictionary to a DataFrame
school_summary_df = pd.DataFrame(school_summary_data)

# Formatting
school_summary_df["Total School Budget"] = school_summary_df["Total School Budget"].map("${:,.2f}".format)
school_summary_df["Per Student Budget"] = school_summary_df["Per Student Budget"].map("${:,.2f}".format)

# Sort the DataFrame by '% Overall Passing' in ascending order
bottom_schools = school_summary_df.sort_values(by='% Overall Passing', ascending=True)

# Display the bottom 5 rows
print("Bottom 5 Schools by % Overall Passing (Ascending Order):")
print(bottom_schools.head())

bottom_schools.head(5)

Bottom 5 Schools by % Overall Passing (Ascending Order):
                       School Type  Total Students Total School Budget  \
school_name                                                              
Hernandez High School   Government            4635       $3,022,020.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Wilson High School     Independent            2283       $1,319,574.00   
Ford High School        Government            2739       $1,763,916.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Hernandez High School            $652.00            68.874865   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Wilson High School               $578.00            69.170828   
Ford High School                 $

School Type  Total Students Total School Budget  \
school_name                                                              
Hernandez High School   Government            4635       $3,022,020.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Wilson High School     Independent            2283       $1,319,574.00   
Ford High School        Government            2739       $1,763,916.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Hernandez High School            $652.00            68.874865   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Wilson High School               $578.00            69.170828   
Ford High School                 $644.00            69.091274   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Hernandez High School              69.186408        80.949299   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Wilson High School                 68.876916        82.785808   
Ford High School                   69.572472        82.438846   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Hernandez High School          81.877023          66.364617  
Huang High School              81.453548          66.712376  
Johnson High School            81.978576          67.191766  
Wilson High School             81.296540          67.455103  
Ford High School               82.219788          67.469880

In [16]:
# Create data series of scores by year levels using conditionals
year_nine = school_data_complete[(school_data_complete["year"] == 9)]
year_ten = school_data_complete[(school_data_complete["year"] == 10)]
year_eleven = school_data_complete[(school_data_complete["year"] == 11)]
year_twelve = school_data_complete[(school_data_complete["year"] == 12)]

# Group each by school name
year_nine_scores = year_nine.groupby('school_name')['maths_score'].mean()
year_ten_scores = year_ten.groupby('school_name')['maths_score'].mean()
year_eleven_scores = year_eleven.groupby('school_name')['maths_score'].mean()
year_twelve_scores = year_twelve.groupby('school_name')['maths_score'].mean()

# Combine series into single DataFrame
maths_scores_by_year = pd.DataFrame({
    '9th Grade': year_nine_scores,
    '10th Grade': year_ten_scores,
    '11th Grade': year_eleven_scores,
    '12th Grade': year_twelve_scores
})

# Minor data wrangling
maths_scores_by_year.index.name = None

# Display the DataFrame
print("Maths Scores by Year Level:")
print(maths_scores_by_year.head(15))

Maths Scores by Year Level:
                       9th Grade  10th Grade  11th Grade  12th Grade
Bailey High School     72.493827   71.897498   72.374900   72.675097
Cabrera High School    72.321970   72.437768   71.008299   70.604712
Figueroa High School   68.477804   68.331586   68.811001   69.325282
Ford High School       69.021609   69.387006   69.248862   68.617811
Griffin High School    72.789731   71.093596   71.692521   71.469178
Hernandez High School  68.586831   68.867156   69.154412   68.985075
Holden High School     70.543307   75.105263   71.640777   73.409639
Huang High School      69.081754   68.533246   69.431345   68.639316
Johnson High School    69.469286   67.990220   68.637730   69.287393
Pena High School       71.996364   72.396000   72.523438   71.187845
Rodriguez High School  71.940722   71.779808   72.364811   72.154626
Shelton High School    72.932075   72.506696   70.097087   72.331536
Thomas High School     69.234273   70.057007   69.657831   69.369822
Wilson

In [17]:
# Create data series of scores by year levels using conditionals
year_nine = school_data_complete[(school_data_complete["year"] == 9)]
year_ten = school_data_complete[(school_data_complete["year"] == 10)]
year_eleven = school_data_complete[(school_data_complete["year"] == 11)]
year_twelve = school_data_complete[(school_data_complete["year"] == 12)]

# Group each by school name
year_nine_scores = year_nine.groupby('school_name')['reading_score'].mean()
year_ten_scores = year_ten.groupby('school_name')['reading_score'].mean()
year_eleven_scores = year_eleven.groupby('school_name')['reading_score'].mean()
year_twelve_scores = year_twelve.groupby('school_name')['reading_score'].mean()

# Combine series into single DataFrame
reading_scores_by_year = pd.DataFrame({
    '9th Grade': year_nine_scores,
    '10th Grade': year_ten_scores,
    '11th Grade': year_eleven_scores,
    '12th Grade': year_twelve_scores
})

# Minor data wrangling
reading_scores_by_year.index.name = None

# Display the DataFrame
print("Reading Scores by Year Level:")
print(reading_scores_by_year.head(15))

Reading Scores by Year Level:
                       9th Grade  10th Grade  11th Grade  12th Grade
Bailey High School     70.901920   70.848265   70.317346   72.195525
Cabrera High School    71.172348   71.328326   71.201245   71.856021
Figueroa High School   70.261682   67.677588   69.152327   69.082126
Ford High School       69.615846   68.988701   70.735964   68.849722
Griffin High School    72.026895   70.746305   72.385042   69.434932
Hernandez High School  68.477569   70.621842   68.418199   69.244136
Holden High School     71.598425   71.096491   73.310680   70.481928
Huang High School      68.670616   69.516297   68.740638   68.671795
Johnson High School    68.719286   69.295029   69.969115   67.992521
Pena High School       70.949091   72.324000   71.703125   71.513812
Rodriguez High School  70.902921   70.137500   71.424453   71.414449
Shelton High School    70.715094   69.879464   71.150485   69.070081
Thomas High School     69.672451   69.741093   70.749398   68.730769
Wils

In [25]:
# Establish the bins
spending_bins = [0, 585, 630, 645, 680]
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [26]:
# Create a copy of the school summary since it has the "Per Student Budget"
#  This step can be skipped but its best to make a copy.
school_spending_df = per_school_summary.copy()

# Display the DataFrame to verify the copy
school_spending_df.head(15)

School Type  Total Students Total School Budget  \
school_name                                                              
Bailey High School      Government            4976       $3,124,928.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Figueroa High School    Government            2949       $1,884,411.00   
Ford High School        Government            2739       $1,763,916.00   
Griffin High School    Independent            1468         $917,500.00   
Hernandez High School   Government            4635       $3,022,020.00   
Holden High School     Independent             427         $248,087.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Pena High School       Independent             962         $585,858.00   
Rodriguez High School   Government            3999       $2,547,363.00   
Shelton High School    Independent            1761       $1,056,600.00   
Thomas High School     Independent            1635       $1,043,130.00   
Wilson High School     Independent            2283       $1,319,574.00   
Wright High School     Independent            1800       $1,049,400.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Bailey High School               $628.00            72.352894   
Cabrera High School              $582.00            71.657158   
Figueroa High School             $639.00            68.698542   
Ford High School                 $644.00            69.091274   
Griffin High School              $625.00            71.788147   
Hernandez High School            $652.00            68.874865   
Holden High School               $581.00            72.583138   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Pena High School                 $609.00            72.088358   
Rodriguez High School            $637.00            72.047762   
Shelton High School              $600.00            72.034072   
Thomas High School               $638.00            69.581651   
Wilson High School               $578.00            69.170828   
Wright High School               $583.00            72.047222   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Bailey High School                 71.008842        91.639871   
Cabrera High School                71.359526        90.850377   
Figueroa High School               69.077993        81.654798   
Ford High School                   69.572472        82.438846   
Griffin High School                71.245232        91.212534   
Hernandez High School              69.186408        80.949299   
Holden High School                 71.660422        89.929742   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Pena High School                   71.613306        91.683992   
Rodriguez High School              70.935984        90.797699   
Shelton High School                70.257808        91.538898   
Thomas High School                 69.768807        83.853211   
Wilson High School                 68.876916        82.785808   
Wright High School                 70.969444        91.777778   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Bailey High School             87.379421          80.084405  
Cabrera High School            89.074273          80.785791  
Figueroa High School           82.807731          67.650051  
Ford High School               82.219788          67.469880  
Griffin High School            88.487738          81.335150  
Hernandez High School          81.877023          66.364617  
Holden High School             88.524590          78.922

In [27]:
# Ensure 'Per Student Budget' column is numeric
school_spending_df['Per Student Budget'] = school_spending_df['Per Student Budget'].str.replace("$", '').astype(float)

# Categorize spending based on bins and labels
school_spending_df["Spending Ranges (Per Student)"] = pd.cut(school_spending_df["Per Student Budget"], bins=spending_bins, labels=labels, right=False)

In [28]:
#  Calculate averages for the desired columns.
spending_maths_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Maths Score"].mean()
spending_reading_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Reading Score"].mean()
spending_passing_maths = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Maths"].mean()
spending_passing_reading = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Reading"].mean()
overall_passing_spending = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Overall Passing"].mean()

# Display the calculated averages
print("Average Maths Scores by Spending Ranges:")
print(spending_maths_scores)
print("\nAverage Reading Scores by Spending Ranges:")
print(spending_reading_scores)
print("\n% Passing Maths by Spending Ranges:")
print(spending_passing_maths)
print("\n% Passing Reading by Spending Ranges:")
print(spending_passing_reading)
print("\nOverall Passing Percentage by Spending Ranges:")
print(overall_passing_spending)

Average Maths Scores by Spending Ranges:
Spending Ranges (Per Student)
<$585       71.364587
$585-630    72.065868
$630-645    69.854807
$645-680    68.884391
Name: Average Maths Score, dtype: float64

Average Reading Scores by Spending Ranges:
Spending Ranges (Per Student)
<$585       70.716577
$585-630    71.031297
$630-645    69.838814
$645-680    69.045403
Name: Average Reading Score, dtype: float64

% Passing Maths by Spending Ranges:
Spending Ranges (Per Student)
<$585       88.835926
$585-630    91.518824
$630-645    84.686139
$645-680    81.568470
Name: % Passing Maths, dtype: float64

% Passing Reading by Spending Ranges:
Spending Ranges (Per Student)
<$585       86.390517
$585-630    87.292423
$630-645    83.763585
$645-680    81.769716
Name: % Passing Reading, dtype: float64

Overall Passing Percentage by Spending Ranges:
Spending Ranges (Per Student)
<$585       76.721458
$585-630    79.876293
$630-645    71.004977
$645-680    66.756253
Name: % Overall Passing, dtype: float

/var/folders/rh/l2gclf9s7654727d9_b6v2th0000gn/T/ipykernel_57329/1709782564.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  spending_maths_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Maths Score"].mean()
/var/folders/rh/l2gclf9s7654727d9_b6v2th0000gn/T/ipykernel_57329/1709782564.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  spending_reading_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Reading Score"].mean()
/var/folders/rh/l2gclf9s7654727d9_b6v2th0000gn/T/ipykernel_57329/1709782564.py:4: FutureWarning: The default of observed=Fal

In [31]:
# Assemble into DataFrame
spending_summary = pd.DataFrame({
    "Average Maths Score": spending_maths_scores,
    "Average Reading Score": spending_reading_scores,
    "% Passing Maths": spending_passing_maths,
    "% Passing Reading": spending_passing_reading,
    "% Overall Passing": overall_passing_spending
})

# Display results
spending_summary.head()

Average Maths Score  Average Reading Score  \
Spending Ranges (Per Student)                                               
<$585                                    71.364587              70.716577   
$585-630                                 72.065868              71.031297   
$630-645                                 69.854807              69.838814   
$645-680                                 68.884391              69.045403   

                               % Passing Maths  % Passing Reading  \
Spending Ranges (Per Student)                                       
<$585                                88.835926          86.390517   
$585-630                             91.518824          87.292423   
$630-645                             84.686139          83.763585   
$645-680                             81.568470          81.769716   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                  76.721458  
$585-630                               79.876293  
$630-645                               71.004977  
$645-680                               66.756253

In [32]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [34]:
# Categorize the school size based on the bins and labels
per_school_summary["School Size"] = pd.cut(per_school_summary["Total Students"], bins=size_bins, labels=group_names)
per_school_summary

School Type  Total Students Total School Budget  \
school_name                                                              
Bailey High School      Government            4976       $3,124,928.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Figueroa High School    Government            2949       $1,884,411.00   
Ford High School        Government            2739       $1,763,916.00   
Griffin High School    Independent            1468         $917,500.00   
Hernandez High School   Government            4635       $3,022,020.00   
Holden High School     Independent             427         $248,087.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Pena High School       Independent             962         $585,858.00   
Rodriguez High School   Government            3999       $2,547,363.00   
Shelton High School    Independent            1761       $1,056,600.00   
Thomas High School     Independent            1635       $1,043,130.00   
Wilson High School     Independent            2283       $1,319,574.00   
Wright High School     Independent            1800       $1,049,400.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Bailey High School               $628.00            72.352894   
Cabrera High School              $582.00            71.657158   
Figueroa High School             $639.00            68.698542   
Ford High School                 $644.00            69.091274   
Griffin High School              $625.00            71.788147   
Hernandez High School            $652.00            68.874865   
Holden High School               $581.00            72.583138   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Pena High School                 $609.00            72.088358   
Rodriguez High School            $637.00            72.047762   
Shelton High School              $600.00            72.034072   
Thomas High School               $638.00            69.581651   
Wilson High School               $578.00            69.170828   
Wright High School               $583.00            72.047222   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Bailey High School                 71.008842        91.639871   
Cabrera High School                71.359526        90.850377   
Figueroa High School               69.077993        81.654798   
Ford High School                   69.572472        82.438846   
Griffin High School                71.245232        91.212534   
Hernandez High School              69.186408        80.949299   
Holden High School                 71.660422        89.929742   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Pena High School                   71.613306        91.683992   
Rodriguez High School              70.935984        90.797699   
Shelton High School                70.257808        91.538898   
Thomas High School                 69.768807        83.853211   
Wilson High School                 68.876916        82.785808   
Wright High School                 70.969444        91.777778   

                       % Passing Reading  % Overall Passing  \
school_name                                                   
Bailey High School             87.379421          80.084405   
Cabrera High School            89.074273          80.785791   
Figueroa High School           82.807731          67.650051   
Ford High School               82.219788          67.469880   
Griffin High School            88.487738          81.335150   
Hernandez High School          81.877023          66.364617   
Holden High School             88.524590        

In [35]:
# Calculate averages for the desired columns.
size_maths_scores = per_school_summary.groupby(["School Size"])["Average Maths Score"].mean()
size_reading_scores = per_school_summary.groupby(["School Size"])["Average Reading Score"].mean()
size_passing_maths = per_school_summary.groupby(["School Size"])["% Passing Maths"].mean()
size_passing_reading = per_school_summary.groupby(["School Size"])["% Passing Reading"].mean()
size_overall_passing = per_school_summary.groupby(["School Size"])["% Overall Passing"].mean()

# Display the calculated averages
print("Average Maths Scores by School Size:")
print(size_maths_scores)
print("\nAverage Reading Scores by School Size:")
print(size_reading_scores)
print("\n% Passing Maths by School Size:")
print(size_passing_maths)
print("\n% Passing Reading by School Size:")
print(size_passing_reading)
print("\nOverall Passing Percentage by School Size:")
print(size_overall_passing)

Average Maths Scores by School Size:
School Size
Small (<1000)         72.335748
Medium (1000-2000)    71.421650
Large (2000-5000)     69.751809
Name: Average Maths Score, dtype: float64

Average Reading Scores by School Size:
School Size
Small (<1000)         71.636864
Medium (1000-2000)    70.720164
Large (2000-5000)     69.576052
Name: Average Reading Score, dtype: float64

% Passing Maths by School Size:
School Size
Small (<1000)         90.806867
Medium (1000-2000)    89.846560
Large (2000-5000)     84.252804
Name: % Passing Maths, dtype: float64

% Passing Reading by School Size:
School Size
Small (<1000)         87.557513
Medium (1000-2000)    86.714149
Large (2000-5000)     83.301185
Name: % Passing Reading, dtype: float64

Overall Passing Percentage by School Size:
School Size
Small (<1000)         79.066348
Medium (1000-2000)    78.039785
Large (2000-5000)     70.293507
Name: % Overall Passing, dtype: float64


/var/folders/rh/l2gclf9s7654727d9_b6v2th0000gn/T/ipykernel_57329/1221066323.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  size_maths_scores = per_school_summary.groupby(["School Size"])["Average Maths Score"].mean()
/var/folders/rh/l2gclf9s7654727d9_b6v2th0000gn/T/ipykernel_57329/1221066323.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  size_reading_scores = per_school_summary.groupby(["School Size"])["Average Reading Score"].mean()
/var/folders/rh/l2gclf9s7654727d9_b6v2th0000gn/T/ipykernel_57329/1221066323.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True

In [36]:
# Assemble into DataFrame
size_summary = pd.DataFrame({
    "Average Maths Score": size_maths_scores,
    "Average Reading Score": size_reading_scores,
    "% Passing Maths": size_passing_maths,
    "% Passing Reading": size_passing_reading,
    "% Overall Passing": size_overall_passing
})

# Display results
size_summary.head()

Average Maths Score  Average Reading Score  \
School Size                                                      
Small (<1000)                 72.335748              71.636864   
Medium (1000-2000)            71.421650              70.720164   
Large (2000-5000)             69.751809              69.576052   

                    % Passing Maths  % Passing Reading  % Overall Passing  
School Size                                                                
Small (<1000)             90.806867          87.557513          79.066348  
Medium (1000-2000)        89.846560          86.714149          78.039785  
Large (2000-5000)         84.252804          83.301185          70.293507

In [37]:
# Create new series using groupby for:
# Type | Average Maths Score | Average Reading Score | % Passing Maths | % Passing Reading | % Overall Passing

type_maths_scores = per_school_summary.groupby(["School Type"])["Average Maths Score"].mean()
type_reading_scores = per_school_summary.groupby(["School Type"])["Average Reading Score"].mean()
type_passing_maths = per_school_summary.groupby(["School Type"])["% Passing Maths"].mean()
type_passing_reading = per_school_summary.groupby(["School Type"])["% Passing Reading"].mean()
type_overall_passing = per_school_summary.groupby(["School Type"])["% Overall Passing"].mean()

In [38]:
# Assemble into DataFrame
type_summary = pd.DataFrame({
    "Average Maths Score": type_maths_scores,
    "Average Reading Score": type_reading_scores,
    "% Passing Maths": type_passing_maths,
    "% Passing Reading": type_passing_reading,
    "% Overall Passing": type_overall_passing
})

# Display results
type_summary.head()

Average Maths Score  Average Reading Score  % Passing Maths  \
School Type                                                                
Government             69.834806              69.675929        84.462375   
Independent            71.368822              70.718933        89.204043   

             % Passing Reading  % Overall Passing  
School Type                                        
Government           83.587562          70.698993  
Independent          86.247789          76.973340

Written Report

Summary of Analysis

-Suprisingly, based on the results for scores by school spending, the higher the budget per student ($645-680 range), the % overall passing is the lowest at 66.76%.

-Based on the results for scores by school size, the smaller the school size (<1000), the higher the % overall passing (79%).

-Based on the scores by school type, independent schoools have higher % overall passing (76.97%) than government schools (70.69%).

-Based on the math score by year level results, the scores stay relatively the same throughout each year level for each school except for Holden High School where I see a 5% additional improvement in Year 10 compared to Year 9. So each school has its stability in its school performance and maintains their scores and not much improvement can be seen throughout each year level.

-Based on the reading score by year level results, the scores stay relatively the same throughout each year level for each school. The biggest improvement shown is 2% improvement at max, for example, Griffin High School in Year 11 but then drops at Year 12 by 3%. So each school has its stability in its school performance and maintains their scores and not much improvement can be seen throughout each year level.

Conclusions/Comparisons from calculations

-The highest total school budget ($3,124,928) in Bailey High School shows the highest % overall passing, 80.08%. However, the second and third hgihest total school budget $3,094,650 in Johnson High School and $3,0222,020 in Hernandez High School show a low % passing score, 67% and 66%, respectively. So perhaps more attention and different implementations would need to be given to Johnson and Hernandez High Schools to improve their % overall passing.

-There is a correlation between school types and % overall passing. The school summary table shows that independent schools tend to have high % overall passing, 78.80% range whilst government schools tend to have low % overall passing in the 66-67% range except for Rodriguez High School, 79%.